dora/core/common/src/main/java/alluxio/membership/ServiceDiscoveryRecipe.java

```diff
@@ -0,0 +1,323 @@
+    try (LockResource lockResource = new LockResource(service.mLock)) {
+      if (service.mLease != null && !mAlluxioEtcdClient.isLeaseExpired(service.mLease)) {
+        LOG.info("Lease attached with service:{} is not expired, bail from here.");
+        return;
+      }
+      String path = service.mServiceEntityName;
+      String fullPath = new StringBuffer().append(mRegisterPathPrefix)
+          .append(MembershipManager.PATH_SEPARATOR)
+          .append(path).toString();
+      try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+        AlluxioEtcdClient.Lease lease = mAlluxioEtcdClient.createLease();
+        Txn txn = mAlluxioEtcdClient.getEtcdClient().getKVClient().txn();
+        ByteSequence keyToPut = ByteSequence.from(fullPath, StandardCharsets.UTF_8);
+        DataOutputStream dos = new DataOutputStream(baos);
+        service.serialize(dos);
+        ByteSequence valToPut = ByteSequence.from(baos.toByteArray());
+        CompletableFuture<TxnResponse> txnResponseFut = txn.If(
+                new Cmp(keyToPut, Cmp.Op.EQUAL, CmpTarget.version(0L)))
+            .Then(Op.put(keyToPut, valToPut, PutOption.newBuilder()
+                .withLeaseId(lease.mLeaseId).build()))
+            .Then(Op.get(keyToPut, GetOption.DEFAULT))
+            .Else(Op.get(keyToPut, GetOption.DEFAULT))
+            .commit();
+        TxnResponse txnResponse = txnResponseFut.get();
+        List<KeyValue> kvs = new ArrayList<>();
+        txnResponse.getGetResponses().stream().map(
+            r -> kvs.addAll(r.getKvs())).collect(Collectors.toList());
+        if (!txnResponse.isSucceeded()) {
+          if (!kvs.isEmpty()) {
+            throw new AlreadyExistsException("Same service kv pair is there but "
+                + "attached lease is expired, this should not happen");
+          }
+          throw new IOException("Failed to new a lease for service:" + service.toString());
+        }
+        Preconditions.checkState(!kvs.isEmpty(), "No such service entry found.");
+        long latestRevision = kvs.stream().mapToLong(kv -> kv.getModRevision())
+            .max().getAsLong();
+        service.mRevision = latestRevision;
+        service.mLease = lease;
+        startHeartBeat(service);
+      } catch (ExecutionException | InterruptedException ex) {
+        throw new IOException("Exception in new-ing lease for service:" + service, ex);
+      }
+    }
+  }
+
+  /**
+   * Register service and start keeping-alive.
+   * @param service
+   * @throws IOException
+   */
+  @GuardedBy("ServiceDiscoveryRecipe#mRegisterLock")
+  public void registerAndStartSync(ServiceEntity service) throws IOException {
+    LOG.info("registering service : {}", service);
+    if (mRegisteredServices.containsKey(service.mServiceEntityName)) {
+      throw new AlreadyExistsException("Service " + service.mServiceEntityName
+          + " already registerd.");
+    }
+    newLeaseInternal(service);
```

Comment:
Not for this PR, but later you might wanna add a timer here and log strangely long attempts.